whisper-cpp: add ROCm and Vulkan support, rename from openai-whisper-cpp

aviallon opened the pull request:
Rewrite to use CMake build
Add aviallon to whisper-cpp's maintainers

Fixes https://github.com/NixOS/nixpkgs/issues/357736  
Fixes https://github.com/NixOS/nixpkgs/issues/349669  
Closes https://github.com/NixOS/nixpkgs/pull/358174

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

aviallon committed:
whisper-cpp: convert to by-name

aviallon committed:
whisper-cpp: add ROCm and Vulkan support, rename from openai-whisper-cpp

Rewrite to use CMake build
Add aviallon to whisper-cpp's maintainers

aviallon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

aviallon commented:
If someone could test on Darwin, it would be awesome.

aviallon commented:
> Great job! I tested it on Darwin, but unfortunately, it didn't work:
> 
> ```
> dyld[98714]: Library not loaded: libwhisper.1.dylib
>   Referenced from: <snip> /nix/store/kxk6zw08rfq61apq09mr5n8z6lndngv8-whisper-cpp-1.7.2/bin/whisper-cpp
>   Reason: tried: 'libwhisper.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibwhisper.1.dylib' (no such file), 'libwhisper.1.dylib' (no such file), '/usr/local/lib/libwhisper.1.dylib' (no such file), '/usr/lib/libwhisper.1.dylib' (no such file, not in dyld cache), '/Users/<snip>/libwhisper.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/<snip>/libwhisper.1.dylib' (no such file), '/Users/<snip>/libwhisper.1.dylib' (no such file), '/usr/local/lib/libwhisper.1.dylib' (no such file), '/usr/lib/libwhisper.1.dylib' (no such file, not in dyld cache)
> [1]    98714 abort      whisper-cpp
> ```

It looks like the RPATH is not set correctly.
It could be because of the fact the binaries are not installation targets, and may need specific patching.
Can you try adding an explicit patchelf rpath fix in postInstall?

I'll also see if I can create a KVM-OSX VM to test the derivation directly.

keysmashes reviewed (commented):
tested Vulkan support with:

```
$ nixpkgs-review pr 365033
$ mkdir /tmp/whisper
$ ./results/whisper-cpp-vulkan-x86_64-linux/bin/whisper-cpp-download-ggml-model base.en /tmp/whisper/
$ ./results/whisper-cpp-vulkan-x86_64-linux/bin/whisper-cpp -m /tmp/whisper/ggml-base.en.bin -m /tmp/whisper/ggml-base.en.bin ~/Downloads/jfk.wav
```

and it seems to work ok:

```
ggml_vulkan: Found 1 Vulkan devices:
ggml_vulkan: 0 = AMD Radeon RX 7900 XTX (RADV NAVI31) ((null)) | uma: 1 | fp16: 1986290034 | warp size: 0
```

with `rocmSupport = true`, it seems to work fine as well (though it takes ~25 minutes to build locally):

```
ggml_cuda_init: found 1 ROCm devices:
  Device 0: AMD Radeon RX 7900 XTX, compute capability 11.0, VMM: no
```

kurnevsky commented on pkgs/by-name/wh/whisper-cpp/package.nix:
```diff
@@ -0,0 +1,185 @@
+      # Build all targets supported by rocBLAS. When updating search for TARGET_LIST_ROCM
+      # in https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/develop/CMakeLists.txt
+      # and select the line that matches the current nixpkgs version of rocBLAS.
+      "-DAMDGPU_TARGETS=${lib.concatStringsSep ";" rocmPackages.clr.gpuTargets}"
```
Could this be moved to args to be able to override it? See https://github.com/ggerganov/llama.cpp/pull/10794

aviallon replied:
Of course. I'll add this.
I also think the ROCm version should be tagged as big parallel.

aviallon replied:
Done

hughobrien commented:
The vanilla build works (and fixes the SIMD issues thank you)
<details>

```
[nix-shell:~/.cache/nixpkgs-review/pr-365033]$ ./results/whisper-cpp-x86_64-linux/bin/whisper-cpp-bench --model "$model"
whisper_init_from_file_with_params_no_state: loading model from '/home/hugh/.local/share/whisper/ggml-medium.en.bin'
whisper_init_with_params_no_state: use gpu    = 1
whisper_init_with_params_no_state: flash attn = 0
whisper_init_with_params_no_state: gpu_device = 0
whisper_init_with_params_no_state: dtw        = 0
whisper_init_with_params_no_state: backends   = 1
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51864
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 1024
whisper_model_load: n_audio_head  = 16
whisper_model_load: n_audio_layer = 24
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 1024
whisper_model_load: n_text_head   = 16
whisper_model_load: n_text_layer  = 24
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 1
whisper_model_load: qntvr         = 0
whisper_model_load: type          = 4 (medium)
whisper_model_load: adding 1607 extra tokens
whisper_model_load: n_langs       = 99
whisper_model_load:      CPU total size =  1533.14 MB
whisper_model_load: model size    = 1533.14 MB
whisper_init_state: kv self size  =   50.33 MB
whisper_init_state: kv cross size =  150.99 MB
whisper_init_state: kv pad  size  =    6.29 MB
whisper_init_state: compute buffer (conv)   =   28.55 MB
whisper_init_state: compute buffer (encode) =  170.15 MB
whisper_init_state: compute buffer (cross)  =    7.72 MB
whisper_init_state: compute buffer (decode) =   98.18 MB

system_info: n_threads = 4 / 16 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | METAL = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | CUDA = 0 | COREML = 0 | OPENVINO = 0 | CANN = 0

whisper_print_timings:     load time =  1724.71 ms
whisper_print_timings:     fallbacks =   0 p /   0 h
whisper_print_timings:      mel time =     0.00 ms
whisper_print_timings:   sample time =     0.00 ms /     1 runs (    0.00 ms per run)
whisper_print_timings:   encode time = 18957.00 ms /     1 runs (18957.00 ms per run)
whisper_print_timings:   decode time =  7771.20 ms /   256 runs (   30.36 ms per run)
whisper_print_timings:   batchd time =  5600.82 ms /   320 runs (   17.50 ms per run)
whisper_print_timings:   prompt time = 62897.61 ms /  4096 runs (   15.36 ms per run)
whisper_print_timings:    total time = 95227.88 ms
```
</details>

However the `vulkan` build did not:

```
[nix-shell:~/.cache/nixpkgs-review/pr-365033]$ ./results/whisper-cpp-vulkan-x86_64-linux/bin/whisper-cpp-bench --model "$model"
whisper_init_from_file_with_params_no_state: loading model from '/home/hugh/.local/share/whisper/ggml-medium.en.bin'
whisper_init_with_params_no_state: use gpu    = 1
whisper_init_with_params_no_state: flash attn = 0
whisper_init_with_params_no_state: gpu_device = 0
whisper_init_with_params_no_state: dtw        = 0
ggml_vulkan: Found 1 Vulkan devices:
Segmentation fault (core dumped)

[nix-shell:~/.cache/nixpkgs-review/pr-365033]$ ./results/whisper-cpp-vulkan-x86_64-linux/bin/whisper-cpp --model "$model" /tmp/out.wav 
whisper_init_from_file_with_params_no_state: loading model from '/home/hugh/.local/share/whisper/ggml-medium.en.bin'
whisper_init_with_params_no_state: use gpu    = 1
whisper_init_with_params_no_state: flash attn = 0
whisper_init_with_params_no_state: gpu_device = 0
whisper_init_with_params_no_state: dtw        = 0
ggml_vulkan: Found 1 Vulkan devices:
Segmentation fault (core dumped)

```

aviallon commented:
@PapayaJackal I tried adding `addAutoPatchelfSearchPath`, to see how it goes on macOS.

rodfersou commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 365033`

---
### `aarch64-darwin`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

rodfersou commented:
```bash
$ nix build --file /nix/store/w1hgvm5rb3rblya8if90cdx5lvwyf4wc-nixpkgs-review-2.12.0/lib/python3.12/site-packages/nixpkgs_review/nix/review-shell.nix --nix-path 'nixpkgs=/Users/rodrigo/.cache/nixpkgs-review/pr-365033-1/nixpkgs nixpkgs-overlays=/var/folders/by/9xt1lyys0qvds8zzbx60c58m0000gn/T/tmppcq_ecwm' --extra-experimental-features 'nix-command no-url-literals' --no-link --keep-going --no-allow-import-from-derivation --argstr local-system aarch64-darwin --argstr nixpkgs-path /Users/rodrigo/.cache/nixpkgs-review/pr-365033-1/nixpkgs --argstr nixpkgs-config-path /var/folders/by/9xt1lyys0qvds8zzbx60c58m0000gn/T/tmp0qmt4adr.nix --argstr attrs-path /Users/rodrigo/.cache/nixpkgs-review/pr-365033-1/attrs.nix
error: builder for '/nix/store/dinzb2x13v9ssxvp36p7m76fk35dc13a-whisper-cpp-1.7.2.drv' failed with exit code 1;
       last 10 log lines:
       >           ^^^^^^^^^^^^^^^
       >   File "/nix/store/whmwqp5qa1lpadz0ik6wfjgjzbbgjh1z-python3-3.12.7-env/lib/python3.12/site-packages/elftools/elf/elffile.py", line 84, in __init__
       >     self._identify_file()
       >   File "/nix/store/whmwqp5qa1lpadz0ik6wfjgjzbbgjh1z-python3-3.12.7-env/lib/python3.12/site-packages/elftools/elf/elffile.py", line 570, in _identify_file
       >     elf_assert(magic == b'\x7fELF', 'Magic number does not match')
       >   File "/nix/store/whmwqp5qa1lpadz0ik6wfjgjzbbgjh1z-python3-3.12.7-env/lib/python3.12/site-packages/elftools/common/utils.py", line 80, in elf_assert
       >     _assert_with_exception(cond, msg, ELFError)
       >   File "/nix/store/whmwqp5qa1lpadz0ik6wfjgjzbbgjh1z-python3-3.12.7-env/lib/python3.12/site-packages/elftools/common/utils.py", line 143, in _assert_with_exception
       >     raise exception_type(msg)
       > elftools.common.exceptions.ELFError: Magic number does not match
       For full logs, run 'nix-store -l /nix/store/dinzb2x13v9ssxvp36p7m76fk35dc13a-whisper-cpp-1.7.2.drv'.
error: builder for '/nix/store/gdcg3r5pwz8m8w0vbf6jk7cy6w76jylc-whisper-cpp-1.7.2.drv' failed with exit code 1;
       last 10 log lines:
       >           ^^^^^^^^^^^^^^^
       >   File "/nix/store/whmwqp5qa1lpadz0ik6wfjgjzbbgjh1z-python3-3.12.7-env/lib/python3.12/site-packages/elftools/elf/elffile.py", line 84, in __init__
       >     self._identify_file()
       >   File "/nix/store/whmwqp5qa1lpadz0ik6wfjgjzbbgjh1z-python3-3.12.7-env/lib/python3.12/site-packages/elftools/elf/elffile.py", line 570, in _identify_file
       >     elf_assert(magic == b'\x7fELF', 'Magic number does not match')
       >   File "/nix/store/whmwqp5qa1lpadz0ik6wfjgjzbbgjh1z-python3-3.12.7-env/lib/python3.12/site-packages/elftools/common/utils.py", line 80, in elf_assert
       >     _assert_with_exception(cond, msg, ELFError)
       >   File "/nix/store/whmwqp5qa1lpadz0ik6wfjgjzbbgjh1z-python3-3.12.7-env/lib/python3.12/site-packages/elftools/common/utils.py", line 143, in _assert_with_exception
       >     raise exception_type(msg)
       > elftools.common.exceptions.ELFError: Magic number does not match
       For full logs, run 'nix-store -l /nix/store/gdcg3r5pwz8m8w0vbf6jk7cy6w76jylc-whisper-cpp-1.7.2.drv'.
error: 2 dependencies of derivation '/nix/store/dasbp7ay07cbiw4l4rp56bxfk5ink0jp-review-shell.drv' failed to build
```

rodfersou commented:
@aviallon, this is the first time I've run those commands, and my environment might also be broken.

Please let me know if there is anything more I can do to help

aviallon commented:
@rodfersou thank you very much.
I think I'll set up a Hackintosh VM in order to debug it.

rodfersou commented:
@aviallon I run the command for full logs looks like the target macos (12) is different from the one I have (11)

`note: 'whisper_encoder_impl' has been marked as being introduced in macOS 12.0 here, but the deployment target is macOS 11.0.0`

```bash
[2/55] Building CXX object src/CMakeFiles/whisper.coreml.dir/coreml/whisper-encoder.mm.o
/tmp/nix-build-whisper-cpp-1.7.2.drv-1/source/src/coreml/whisper-encoder.mm:31:44: warning: 'whisper_encoder_impl' is only available on macOS 12.0 or newer [-
Wunguarded-availability-new]
    const void * data = CFBridgingRetain([[whisper_encoder_impl alloc] initWithContentsOfURL:url_model configuration:config error:nil]);
                                           ^~~~~~~~~~~~~~~~~~~~
/tmp/nix-build-whisper-cpp-1.7.2.drv-1/source/src/coreml/whisper-encoder-impl.h:41:12: note: 'whisper_encoder_impl' has been marked as being introduced in macOS 12.0 here, but the deployment target is macOS 11.0.0
@interface whisper_encoder_impl : NSObject
           ^
/tmp/nix-build-whisper-cpp-1.7.2.drv-1/source/src/coreml/whisper-encoder.mm:31:44: note: enclose 'whisper_encoder_impl' in an @available check to silence this warning
    const void * data = CFBridgingRetain([[whisper_encoder_impl alloc] initWithContentsOfURL:url_model configuration:config error:nil]);
                                           ^~~~~~~~~~~~~~~~~~~~
/tmp/nix-build-whisper-cpp-1.7.2.drv-1/source/src/coreml/whisper-encoder.mm:65:9: warning: 'whisper_encoder_implOutput' is only available on macOS 12.0 or newer [-Wunguarded-availability-new]
        whisper_encoder_implOutput * outCoreML = [(__bridge id) ctx->data predictionFromLogmel_data:inMultiArray error:nil];
        ^~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/nix-build-whisper-cpp-1.7.2.drv-1/source/src/coreml/whisper-encoder-impl.h:29:12: note: 'whisper_encoder_implOutput' has been marked as being introduced in macOS 12.0 here, but the deployment target is macOS 11.0.0
@interface whisper_encoder_implOutput : NSObject<MLFeatureProvider>
           ^
/tmp/nix-build-whisper-cpp-1.7.2.drv-1/source/src/coreml/whisper-encoder.mm:65:9: note: enclose 'whisper_encoder_implOutput' in an @available check to silence this warning
        whisper_encoder_implOutput * outCoreML = [(__bridge id) ctx->data predictionFromLogmel_data:inMultiArray error:nil];
        ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

rodfersou commented:
@aviallon can we try with SDK12 instead?

rodfersou commented:
this is what happens when try to run the failed binary

```bash
 ...failed_results/whisper-cpp-aarch64-darwin/bin 
$ ./whisper-cpp
dyld[10672]: Library not loaded: libwhisper.1.dylib
  Referenced from: <3DC12AEA-5D8F-3F2A-8F4F-964BFAD19413> /nix/store/xp339mpr218ppnkm6fcrr9j3c6ckhx09-whisper-cpp-1.7.2/bin/whisper-cpp
  Reason: tried: 'libwhisper.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibwhisper.1.dylib' (no such file), 'libwhisper.1.dylib' (no such file), '/usr/local/lib/libwhisper.1.dylib' (no such file), '/usr/lib/libwhisper.1.dylib' (no such file, not in dyld cache), '/nix/store/xp339mpr218ppnkm6fcrr9j3c6ckhx09-whisper-cpp-1.7.2/bin/libwhisper.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/nix/store/xp339mpr218ppnkm6fcrr9j3c6ckhx09-whisper-cpp-1.7.2/bin/libwhisper.1.dylib' (no such file), '/nix/store/xp339mpr218ppnkm6fcrr9j3c6ckhx09-whisper-cpp-1.7.2/bin/libwhisper.1.dylib' (no such file), '/usr/local/lib/libwhisper.1.dylib' (no such file), '/usr/lib/libwhisper.1.dylib' (no such file, not in dyld cache)
[1]    10672 abort      ./whisper-cpp
```

rodfersou commented:
follow the full logs

[whisper-cpp-aarch64-darwin.log](https://github.com/user-attachments/files/18220857/whisper-cpp-aarch64-darwin.log)
[whisper-cpp-vulkan-aarch64-darwin.log](https://github.com/user-attachments/files/18220858/whisper-cpp-vulkan-aarch64-darwin.log)

rodfersou commented:
@aviallon this should help https://stackoverflow.com/a/11585225

rodfersou commented on pkgs/by-name/wh/whisper-cpp/package.nix:
```diff
@@ -0,0 +1,192 @@
+    ]
+    ++ optionals metalSupport [
+      (cmakeFeature "CMAKE_C_FLAGS" "-D__ARM_FEATURE_DOTPROD=1")
+      (cmakeBool "LLAMA_METAL_EMBED_LIBRARY" true)
```
```suggestion
      (cmakeBool "WHISPER_METAL_EMBED_LIBRARY" true)
```

rodfersou replied:
@aviallon this should fix for OSX

rodfersou replied:
the original code also had few more flags:

```nix
  env = lib.optionalAttrs stdenv.hostPlatform.isDarwin {
    WHISPER_COREML = "1";
    WHISPER_COREML_ALLOW_FALLBACK = "1";
    WHISPER_METAL_EMBED_LIBRARY = "1";
  }
```

lucasewd commented:
# [nixpkgs-reviewd](https://github.com/nixosbrasil/nixpkgs-reviewd) workflow started

**Status and logs**: https://github.com/nixosbrasil/nixpkgs-reviewd/actions/runs/12474647914

lucasewd commented:
# [nixpkgs-reviewd](https://github.com/nixosbrasil/nixpkgs-reviewd) workflow finished

**Status and logs**: https://github.com/nixosbrasil/nixpkgs-reviewd/actions/runs/12474647914


## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 365033`

---
### `x86_64-darwin`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 365033`

---
### `aarch64-darwin`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 365033`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

hughobrien commented:
Looking forward to this landing, please let us know if we can assist.

lucasew commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 365033`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 8 packages built:</summary>
  <ul>
    <li>kodiPackages.iagl</li>
    <li>kodiPackages.youtube</li>
    <li>netbird-dashboard</li>
    <li>nixos-firewall-tool</li>
    <li>nzbhydra2</li>
    <li>qownnotes</li>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

DontEatOreo commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 365033`

---
### `x86_64-darwin`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

aviallon commented:
Quite frankly, I can't fix macOS support.
It is probably best if I mark it broken, and a macOS user spends the time troubleshooting this.

DontEatOreo commented:
I daily drive an `aarch64-darwin` system and I'm interested in `whsper-cpp`, so I'll give it a try. Python and C++ are kind of outside my scope, but if I figure out a fix, I'll make a PR

I suppose it's best to mark it broken for now to not delay this PR any further

While this PR is not merged, I think it will be good idea to bump version to `1.7.4`

rodfersou commented:
@aviallon did you try my suggestions? I'm pretty sure adding the environment variables I told should fix..

rodfersou reviewed (changes requested):
Please add the environment variables to make it work for OSX, as requested in the review.

aviallon commented:
@rodfersou I can try, but it is unlikely to make any difference, as the issue is caused by a wrong RPATH due to the binaries not being part of CMake's install target.
The environment variables are no longer used by the project, per CMake's warnings and the project's source code.

rodfersou commented:
did you notice the typo that you keep LLAMA instead of WHISPER in the environment variable?

aviallon commented:
@rodfersou @lucasew I reworked the PR to patch `CMakeLists.txt` instead, which means we now have proper install targets, and proper `RPATH`.
I also added a trivial `installCheckPhase`, to make sure the built binary doesn't crash outright.

aviallon commented:
@rodfersou According to https://logs.ofborg.org/?attempt_id=d3509326-6aa2-439d-a4a1-7388ad8189a1&key=nixos%2Fnixpkgs.365033, it doesn't even build for x86_64-darwin...

```
FAILED: src/CMakeFiles/whisper.coreml.dir/coreml/whisper-encoder.mm.o 
/nix/store/raqax5qp65an4jcwj1bzd5lq589ywpg0-clang-wrapper-16.0.6/bin/clang++ -Dwhisper_coreml_EXPORTS -I/tmp/nix-build-whisper-cpp-1.7.2.drv-0/source/src/. -F/nix/store/6fb3yzhx0dlhzjk8s1mwkmn505yqbmmb-apple-sdk-10.12.2/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks -O3 -DNDEBUG -isysroot /nix/store/6fb3yzhx0dlhzjk8s1mwkmn505yqbmmb-apple-sdk-10.12.2/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -fPIC -fobjc-arc -MD -MT src/CMakeFiles/whisper.coreml.dir/coreml/whisper-encoder.mm.o -MF src/CMakeFiles/whisper.coreml.dir/coreml/whisper-encoder.mm.o.d -o src/CMakeFiles/whisper.coreml.dir/coreml/whisper-encoder.mm.o -c /tmp/nix-build-whisper-cpp-1.7.2.drv-0/source/src/coreml/whisper-encoder.mm
In file included from /tmp/nix-build-whisper-cpp-1.7.2.drv-0/source/src/coreml/whisper-encoder.mm:6:
/tmp/nix-build-whisper-cpp-1.7.2.drv-0/source/src/coreml/whisper-encoder-impl.h:8:9: fatal error: 'CoreML/CoreML.h' file not found
#import <CoreML/CoreML.h>
        ^~~~~~~~~~~~~~~~~
1 error generated.
ninja: build stopped: subcommand failed.
```

aviallon commented:
I tried disabling CoreML support by default, and marking the package broken when CoreML is enabled.

rodfersou commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 365033`

rodfersou commented:
@aviallon I'm confused, not sure if it's working

aviallon commented:
@ofborg build whisper-cpp

aviallon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 365033`

aviallon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 365033`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

aviallon commented:
@rodfersou there was a very weird bug there, I re-created my branch on top of master, which should fix it.

rodfersou commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 365033`

rodfersou commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 365033`

aviallon commented:
Your nixpkgs-review built nothing at all?!
Perhaps you should specify it which package you want to have built, explicitly.

rodfersou commented:
can you guide me how to run the command?

before the last changes was at least trying to build, now it's not anymore

lucasew commented:
> can you guide me how to run the command?
> 
> before the last changes was at least trying to build, now it's not anymore

Try with `--eval local`

rodfersou commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 365033`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>whisper-cpp</li>
    <li>whisper-cpp-vulkan</li>
  </ul>
</details>

rodfersou commented:
@lucasew it works!!! thanks

aviallon commented:
@rodfersou Hurra!!!

aviallon commented:
Quite frankly, if it works with aarch64-darwin, x86_64-linux and aarch64-linux, there is a 97.9% chance it'll work on x86_64-linux.
If someone with write access could review and merge, it'd be welcomed!

hughobrien commented:
```
[nix-shell:~/.cache/nixpkgs-review/pr-365033-2/results]$ hyperfine "./whisper-cpp-x86_64-linux/bin/whisper-cpp --threads $(nproc) --language en --model ~/.local/share/whisper/ggml-medium.en.bin --file /data/whisper-test.wav"

Benchmark 1: ./whisper-cpp-x86_64-linux/bin/whisper-cpp --threads 16 --language en --model ~/.local/share/whisper/ggml-medium.en.bin --file /data/whisper-test.wav
  Time (mean ± σ):     122.031 s ±  0.253 s    [User: 1545.379 s, System: 51.247 s]
  Range (min … max):   121.624 s … 122.399 s    10 runs

```

The previous version with force-enabled CPU vector extensions.
```
hyperfine "whisper-cpp --threads $(nproc) --language en --model ~/.local/share/whisper/ggml-medium.en.bin --file /data/whisper-test.wav"
Benchmark 1: whisper-cpp --threads 16 --language en --model ~/.local/share/whisper/ggml-medium.en.bin --file /data/whisper-test.wav
  Time (mean ± σ):     136.110 s ±  4.581 s    [User: 2066.905 s, System: 6.383 s]
  Range (min … max):   129.642 s … 146.108 s    10 runs
```

hughobrien commented:
However the vulkan build is still segfaulting:

```
[nix-shell:~/.cache/nixpkgs-review/pr-365033-2/results]$ ./whisper-cpp-x86_64-linux/bin/whisper-cpp-bench --model ~/.local/share/whisper/ggml-medium.en.bin
whisper_init_from_file_with_params_no_state: loading model from '/home/hugh/.local/share/whisper/ggml-medium.en.bin'
whisper_init_with_params_no_state: use gpu    = 1
whisper_init_with_params_no_state: flash attn = 0
whisper_init_with_params_no_state: gpu_device = 0
whisper_init_with_params_no_state: dtw        = 0
whisper_init_with_params_no_state: backends   = 1
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51864
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 1024
whisper_model_load: n_audio_head  = 16
whisper_model_load: n_audio_layer = 24
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 1024
whisper_model_load: n_text_head   = 16
whisper_model_load: n_text_layer  = 24
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 1
whisper_model_load: qntvr         = 0
whisper_model_load: type          = 4 (medium)
whisper_model_load: adding 1607 extra tokens
whisper_model_load: n_langs       = 99
whisper_model_load:      CPU total size =  1533.14 MB
whisper_model_load: model size    = 1533.14 MB
whisper_init_state: kv self size  =   50.33 MB
whisper_init_state: kv cross size =  150.99 MB
whisper_init_state: kv pad  size  =    6.29 MB
whisper_init_state: compute buffer (conv)   =   28.55 MB
whisper_init_state: compute buffer (encode) =  170.15 MB
whisper_init_state: compute buffer (cross)  =    7.72 MB
whisper_init_state: compute buffer (decode) =   98.18 MB

system_info: n_threads = 4 / 16 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | METAL = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | CUDA = 0 | COREML = 0 | OPENVINO = 0 | CANN = 0

whisper_print_timings:     load time =  1027.43 ms
whisper_print_timings:     fallbacks =   0 p /   0 h
whisper_print_timings:      mel time =     0.00 ms
whisper_print_timings:   sample time =     0.00 ms /     1 runs (    0.00 ms per run)
whisper_print_timings:   encode time = 19500.96 ms /     1 runs (19500.96 ms per run)
whisper_print_timings:   decode time = 10304.86 ms /   256 runs (   40.25 ms per run)
whisper_print_timings:   batchd time =  6413.53 ms /   320 runs (   20.04 ms per run)
whisper_print_timings:   prompt time = 61788.45 ms /  4096 runs (   15.09 ms per run)
whisper_print_timings:    total time = 98009.29 ms
```

```
[nix-shell:~/.cache/nixpkgs-review/pr-365033-2/results]$ ./whisper-cpp-vulkan-x86_64-linux/bin/whisper-cpp-bench --model ~/.local/share/whisper/ggml-medium.en.bin
whisper_init_from_file_with_params_no_state: loading model from '/home/hugh/.local/share/whisper/ggml-medium.en.bin'
whisper_init_with_params_no_state: use gpu    = 1
whisper_init_with_params_no_state: flash attn = 0
whisper_init_with_params_no_state: gpu_device = 0
whisper_init_with_params_no_state: dtw        = 0
ggml_vulkan: Found 1 Vulkan devices:
Segmentation fault (core dumped)
```

aviallon commented:
@hughobrien your issue is probably a hardware or a driver bug.
Perhaps you could try to debug it, get a good stacktrace, and eventually post it here or on mesa.

hughobrien commented:
I tried on a system with a dGPU and it was successful

```
[nix-shell:~/.cache/nixpkgs-review/pr-365033]$ hyperfine "./results/whisper-cpp-x86_64-linux/bin/whisper-cpp --threads $(nproc) --language en --model /data/ggml-medium.en.bin --file /tmp/whisper-test.wav"
Benchmark 1: ./results/whisper-cpp-x86_64-linux/bin/whisper-cpp --threads 16 --language en --model /data/ggml-medium.en.bin --file /tmp/whisper-test.wav
  Time (mean ± σ):     41.916 s ±  0.881 s    [User: 644.525 s, System: 3.312 s]
  Range (min … max):   39.779 s … 42.556 s    10 runs
```

```
[nix-shell:~/.cache/nixpkgs-review/pr-365033]$ hyperfine "./results/whisper-cpp-vulkan-x86_64-linux/bin/whisper-cpp --threads $(nproc) --language en --model /data/ggml-medium.en.bin --fi
le /tmp/whisper-test.wav"
Benchmark 1: ./results/whisper-cpp-vulkan-x86_64-linux/bin/whisper-cpp --threads 16 --language en --model /data/ggml-medium.en.bin --file /tmp/whisper-test.wav
  Time (mean ± σ):     31.045 s ±  3.903 s    [User: 2.182 s, System: 0.471 s]
  Range (min … max):   29.737 s … 42.153 s    10 runs
```

hughobrien commented:
I'm testing rocm support now.

Can you speak to the logic behind exposing `whisper-cpp-vulkan` but not `whisper-cpp-rocm` or `-cuda`?

aviallon commented:
> I'm testing rocm support now.
> 
> Can you speak to the logic behind exposing `whisper-cpp-vulkan` as a top level package but not `whisper-cpp-rocm` or `-cuda`?
> 
> Edit: I assume I'm doing something wrong, but do you have steps/guidelines for how to run the `rocm` version?
> Second Edit: What I appear to be hitting is a build issue with `rocm-llvm-libcxx`, (x86_64 nixos-unstable). Obviously that's out of scope for this PR but I'd prefer we have working examples / demos of how to use the different artifacts before we merge.
> 
> Naturally the upstream project is fast-moving and we don't need to do their documentation for them, but a few one-liner examples I think would go a long way.
> 
> 
> Third edit:
> Some jury-rigging later and I got this  result
> ```
> hyperfine "whisper-cpp --threads $(nproc) --language en --model /data/ggml-medium.en.bin --file /tmp/whisper-test.wav"
> Benchmark 1: whisper-cpp --threads 16 --language en --model /data/ggml-medium.en.bin --file /tmp/whisper-test.wav
>   Time (mean ± σ):     10.198 s ±  0.053 s    [User: 10.278 s, System: 2.217 s]
>   Range (min … max):   10.111 s … 10.281 s    10 runs
> ```
> 
> Same machine as the vulkan test above, but 3x faster. Is that expected?
> 
> <details>
> 
> ```
> whisper_init_from_file_with_params_no_state: loading model from '/data/ggml-medium.en.bin'                                    
> whisper_init_with_params_no_state: use gpu    = 1                                                                                                  
> whisper_init_with_params_no_state: flash attn = 0                                                                                                  
> whisper_init_with_params_no_state: gpu_device = 0    
> whisper_init_with_params_no_state: dtw        = 0                        
> ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no
> ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no 
> ggml_cuda_init: found 1 ROCm devices:                                                                                                                                                                                                                                                                  
>   Device 0: AMD Radeon RX 7900 GRE, compute capability 11.0, VMM: no
> whisper_init_with_params_no_state: backends   = 2
> whisper_model_load: loading model                                        
> whisper_model_load: n_vocab       = 51864
> whisper_model_load: n_audio_ctx   = 1500
> whisper_model_load: n_audio_state = 1024
> whisper_model_load: n_audio_head  = 16
> whisper_model_load: n_audio_layer = 24
> whisper_model_load: n_text_ctx    = 448
> whisper_model_load: n_text_state  = 1024
> whisper_model_load: n_text_head   = 16
> whisper_model_load: n_text_layer  = 24
> whisper_model_load: n_mels        = 80
> whisper_model_load: ftype         = 1
> whisper_model_load: qntvr         = 0
> whisper_model_load: type          = 4 (medium)
> whisper_model_load: adding 1607 extra tokens
> whisper_model_load: n_langs       = 99
> whisper_model_load:    ROCm0 total size =  1533.14 MB
> whisper_model_load: model size    = 1533.14 MB
> whisper_backend_init_gpu: using CUDA backend
> whisper_init_state: kv self size  =   50.33 MB
> whisper_init_state: kv cross size =  150.99 MB
> whisper_init_state: kv pad  size  =    6.29 MB
> whisper_init_state: compute buffer (conv)   =   29.51 MB
> whisper_init_state: compute buffer (encode) =  170.15 MB
> whisper_init_state: compute buffer (cross)  =    7.72 MB
> whisper_init_state: compute buffer (decode) =   99.10 MB
> 
> system_info: n_threads = 16 / 16 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | METAL = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | CUDA = 1 | COREML = 0 | OPENVINO = 0 | CANN = 0
> 
> main: processing '/tmp/whisper-test.wav' (1920000 samples, 120.0 sec), 16 threads, 1 processors, 5 beams + best of 5, lang = en, task = transcribe, timestamps = 1 ...
> ```
> 
> </detaisl>

Yes, it is expected. The vulkan backend is immature, and some ops require to fallback to CPU. Which is very very slow.

hughobrien commented:
Approved, thank you for your work here!

skissue reviewed (approved):
Vulkan works flawlessly on my machine :tada: (x86-64 Linux, RX 7800 XT).

GaetanLepage reviewed (approved):
LGTM, well done !
